### ISSUE TITLE
REQUEST: key/value storage for the app

### ISSUE THREAD

**Issue by drwasho** [Jasonette/JASONETTE-iOS#128]
I've discussed this with @gliechtenstein and I'm not sure if there's another issue like this, but essentially it would be nice if there was a place to store data in the app (e.g. `$storage`).

There are a couple of use-cases:

1. **Credential management.** Storage of credentials for decentralized apps (where a remote server will have a different IP address etc per user, so I can't use `$keys`). This prevents the user from having to add their ip address, port, username, and password every time they open the app. Also allows me to store more than one credential if the user is managing multiple servers.
2. **Performance optimisation.** Some data shared between different views doesn't need a new network call to access it. The data can be passed via `$params`, but this can get cumbersome in complex apps.

Additionally the data stored in `$storage` needs to be accessible to authorised views to prevent malicious views trying to scoop up data from any data in `$storage`.

**Comment by steve21124** [comment]
Both these using levelDB
iOS version
https://github.com/matehat/Objective-LevelDB
Android version
https://github.com/nhachicha/SnappyDB

**Comment by Joshfindit** [comment]
If we're considering db backends, how about [Realm](https://realm.io/)?

iOS demo app: https://realm.io/docs/get-started/cocoa-demo-app/
Android demo app: https://realm.io/docs/get-started/android-demo-app/

**Comment by steve21124** [comment]
fyi: https://github.com/realm/realm-cocoa/issues/3242
Realm did not support dynamic schema

**Comment by Joshfindit** [comment]
@steve21124 I wasn't aware of that. It would still work for key/value if both are strings, but unless they start supporting dynamic schema, it may not grow past that.

**Comment by Joshfindit** [comment]
Edit: Not sure how this fits in since it relates to the Realm Mobile Platform, but [realm.io](https://realm.io/news/alexander-stigsen-introducing-the-realm-mobile-platform/?utm_source=news-list&utm_medium=email&utm_content=rmdb-content) states

"You can actually change the scheme on the fly, you can push data to devices, and with the event frame you can introduce to an existing infrastructure."

**Comment by steve21124** [comment]
fyi: I have been using realm for awhile and satisfy with it even though it has its limitation. Overall, i also think better to have realm compare to other since it is open source and commercial and very active community

**Comment by drwasho** [comment]
I don't think we need anything that complicated, just a basic key-value storage would do.

The real problem is in terms of security, and ensuring that other view can't access information that is saved by a certain view.

For example, let's say 'View A' saves `username` and `password`; 'View B' shouldn't be able to access that data. Next, let's say that the author of 'View A' creates 'View C' and wants to access data created by 'View A'.

This is the sort of basic functionality that needs to be supported. The problem is figuring out a data structure to make that work.

**Comment by gliechtenstein** [comment]
I finally have a couple of ideas on this, and posted an issue on this, please check it out. Feedback appreciated: https://github.com/Jasonette/documentation/issues/56
